### PR TITLE
Only link to gperftools in profiling and opt modes

### LINKED ISSUE #17490: Only link to gperftools in profiling modes
## Reason
gperftools hijacks functions like `malloc` which destroys the ability to get any usable information out of valgrind. We should save some modes (`devel` is a great candidate because it has debugging symbols and some level of optimization (`-O2`)) for running with valgrind.

## Design
Edit `build.mk` to only append flags in profiling modes.

## Impact
Make valgrind usable for some mode builds.

### PR DESCRIPTION
**Pull request by lindsayad**
Closes #17490

### CODE REVIEW

**Comment by rwcarlsen** [comment]
The only downside to this is that if you want to use google's faster `malloc` implementation in these libraries, you can't just set GPERF_DIR to use it.  But - yeah - this seems reasonable.

**Comment by lindsayad** [comment]
I could add back `opt` (although I admittedly will have to google 'or' operations for `Makefiles` :rofl: )

**Comment by fdkong** [comment]
So we can not use dbg or opt with gperf tools anymore ?

**Comment by lindsayad** [comment]
> So we can not use dbg or opt with gperf tools anymore ?

I don't know why you would ever want to do profiling with a `dbg` mode build. You want to test the speed of Roy's and my assertions? :smile: I personally don't ever profile with `opt` because we don't pass `-g` and so I can't look at lines numbers. But I suppose that you do run profiling with `opt` just to see stack traces which is reasonable. I suppose I will add back `opt`. I am fine with that.

**Comment by fdkong** [comment]
> I suppose I will add back opt. I am fine with that.

Thanks. I do use "opt" with the profiling tools. Right, I do not think I will profile "dbg" since nothing can be slower than "dbg" ;-)

File: framework/build.mk [hunk 1]
```diff
@@ -46,9 +46,15 @@ ifneq (x$(MOOSE_NO_PERF_GRAPH), x)
 endif
 
 ifneq ($(GPERF_DIR), )
+ifneq (,$(findstring prof,$(METHOD)))
     libmesh_CXXFLAGS += -DHAVE_GPERFTOOLS -I$(GPERF_DIR)/include
     libmesh_LDFLAGS := -L$(GPERF_DIR)/lib -ltcmalloc_and_profiler $(libmesh_LDFLAGS)
 endif
+ifeq (opt,$(METHOD))
```
Comment (fdkong) [inline framework/build.mk:53]: The following two lines are the same as the above two lines. 

Is it possible to combine them? Introduce an `or` operator?

**Review by fdkong — CHANGES_REQUESTED** [review]
Not sure if we can combine the duplicated code?

File: framework/build.mk (hunk 1 above)
Comment (lindsayad) [reply framework/build.mk:53]: I am no `Makefile` expert. This was the best stackoverflow post I found but it doesn't look great to me and it's more simple than our case: https://stackoverflow.com/questions/7656425/makefile-ifeq-logical-or

Comment (YaqiWang) [reply framework/build.mk:53]: Better to issue an error in other modes to let users know they should not expect gperftools to work in those modes.

Comment (rwcarlsen) [reply framework/build.mk:53]: Yeah - we should document this behavior in the profiling docs - that perf tools/tcmalloc are not linked when building in dbg/devel mode regardless of whether gperf_dir is set.

Comment (lindsayad) [reply framework/build.mk]: Ok I removed duplication, added an error, and updated profiling documentation

**Review by fdkong — APPROVED** [review]
Now, it looks much better to me

**Pull request merged**